VSock: Initial datagram implementation for Argo.

eric-ch opened the pull request:
### An out-of-tree module for VSock DGRAM over Argo.

Based on headers released with Xen 4.12.
A copy of Xen `argo.h` header is added to the out-of-tree module to ease the compilation process against existing kernel headers, since it is not shipped as part of Linux headers.
Also re-introduce `hypercall5` in `argo-compat.h`, removed from upstream as it was no longer used, see comments in the header.

### Add a test utility for VSock/Argo DGRAM transfers

Simple test tool to send/recv things over VSock through Argo and ease testing. It tries to mimic `netcat` usage to not look too unfamiliar. `socat` should be able to achieve the same, with a steeper learning curve. This will be made easier once STREAM communication are implemented I would think.

### Provide instructions in a README

Add some instruction on how to test VSock/Argo on existing Xen 4.12.y installations. Try to address caveats that may be encountered and provide some details as to what is this all about.

dozylynx commented on vsock-argo/module/argo_ring.c:
```diff
@@ -0,0 +1,409 @@
+/*
+ * Initialisation and cleanup of the VIRQ.
+ */
+static int argo_virq = 11;
```
any reason not to use VIRQ_ARGO instead of bare 11?

eric-ch replied:
None, it should be VIRQ_ARGO.

dozylynx commented on vsock-argo/module/argo_ring.c:
```diff
@@ -0,0 +1,409 @@
+	if (rc < 0)
+		return rc;
+
+	argo_virq = rc;
```
I think this is an _IRQ_ value (so should be named `argo_irq` instead), not a _VIRQ_ -- or am I misreading/misunderstanding what `bind_virq_to_irqhandler` returns?

eric-ch replied:
You are correct, this is also a discrepancy as `bind_virq_to_irqhandler` should take `ARGO_VIRQ` and not `argo_virq` as first argument.

dozylynx commented on vsock-argo/module/argo_ring.h:
```diff
@@ -0,0 +1,68 @@
+ * Arbitrary fixed ring size for now.
+ * TODO: Should be configurable through sysfs?
+ */
+static const size_t ring_len = 4 * PAGE_SIZE;
```
The previous Linux Argo and v4v drivers have a default ring size of 32 pages -- I think preserving that default value makes sense.

eric-ch replied:
Agreed, this would also give some background for the choice.

eric-ch commented on vsock-argo/README.md:
```diff
@@ -0,0 +1,130 @@
+### Test client
+
+A test `netcat` like utility is provided to simplify send and receive operation
+using VSock datagram over Argo, the tool is call `cargo` and roughly follows
```
Absolutely, I merely named it "cargo" as it seemed fitting for packets sent on a ship, I should have dig a little deeper to find the name conflict.

eric-ch commented on vsock-argo/README.md:
```diff
@@ -0,0 +1,130 @@
+
+This software is still under developpment and should be considered
+experimental.
+
```
Correct, I only tested this with x86 64 bit guests.
I will add that to the comment unless I find an ARM setup in the mean time. I should be able to test x86 32 bit guests though.

dozylynx replied:
gentle ping on adding a x86-64 comment here; can be removed later when tested elsewhere

eric-ch commented on vsock-argo/module/include/xen/argo.h:
```diff
@@ -0,0 +1,290 @@
+/******************************************************************************
```
Two changes were necessary to build the module out-of-tree:

### include "xen.h"

```diff
+#ifdef __XEN__
 #include "xen.h"
+#else
+#include <xen/interface/xen.h>
+#endif
```
Fix compilation outside of Xen tree using this header:
```
 In file included from <tree>/linux-xen-argo/vsock-argo/module/argo_ring.h:6,
                  from <tree>/linux-xen-argo/vsock-argo/module/argo_transport.c:14:
 <tree>/linux-xen-argo/vsock-argo/module/include/xen/argo.h:32:10: fatal error: xen.h: No such file or directory
  #include "xen.h"
           ^~~~~~~
```

### definition of XEN_GUEST_HANDLE

```diff
+#ifdef DEFINE_XEN_GUEST_HANDLE
+DEFINE_XEN_GUEST_HANDLE(uint8_t);
+#endif
+

+#ifdef XEN_GUEST_HANDLE_64
+    XEN_GUEST_HANDLE_64(uint8_t) iov_hnd;
+#else
+    uint64_t iov_hnd;
+#endif
-    XEN_GUEST_HANDLE(uint8) iov_hnd;

```

```
In file included from <tree>/linux-xen-argo/vsock-argo/module/argo_ring.h:6,
                 from <tree>/linux-xen-argo/vsock-argo/module/argo_transport.c:14:
<tree>/linux-xen-argo/vsock-argo/module/include/xen/argo.h:60:5: error: expected specifier-qualifier-list before ‘XEN_GUEST_HANDLE’
     XEN_GUEST_HANDLE(uint8) iov_hnd;
     ^~~~~~~~~~~~~~~~
```

---

Using the public Xen header might not be the best thing here. Creating a new one is risky though, since there is no ABI version to check against if Argo declaration change at some point. Introducing one would have to be done upstream and add a new `argo_op`.

dozylynx replied:
The two changes you indicate above are fine with me. I've seen the same in local development. The result of applying those changes to the upstream header produces the header we want for this now.

eric-ch commented:
### v2
* Fixed reported typo
* Renamed `cargo` to `hatch` (after-all the program is only the thing through which the cargo is sent)
* Refreshed `argo.h` header to match latest release. Still the two hunks documented above are left to attend out-of-tree module compilation (see https://github.com/OpenXT/linux-xen-argo/pull/1#discussion_r284781031)
* Renamed incorrectly named `argo_virq` variable to `argo_irq` and use `ARGO_VIRQ` in `bind_virq_to_irqhandler` as it should have been
* Set default `ring_len` to 32 pages to match previous implementations default.

eric-ch commented:
### v3
- Fixed typo in `argo_ring_alloc`, sanity checks are to be done against argument `len`, not the global `ring_len`.

dozylynx commented on vsock-argo/module/argo_ring.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+static inline size_t argo_ring_has_data_no_wrap(const struct argo_ring_hnd *h)
+{
```
Is this function a duplicate implementation? See lines 195-205 above, where there's an identical looking function with a slightly different name (no underscore in `no_wrap`).

eric-ch replied:
It definitely is, the other one is not even used... My bad.

dozylynx commented on vsock-argo/module/argo-compat.h:
```diff
@@ -0,0 +1,43 @@
+{
+	int ret;
+
+	stac();
```
This header won't work for ARM: `stac` and `clac` aren't available, and I expect that the `_hypercall5` implementation will differ; so: maybe this header belongs in an asm-x86 folder?

eric-ch replied:
Yes, absolutely! With a condition for `ccflags` in `Kbuild` against `CONFIG_X86=y`.

dozylynx commented on vsock-argo/module/argo_ring.c:
```diff
@@ -0,0 +1,409 @@
+}
+
+static void argo_gfn_array_free(struct argo_gfn_array *ga)
+{
```
any reason not to drop this function and just `kfree` instead at its call sites? it's extra kernel code being added here

eric-ch replied:
Just a habit of providing a `_free` for an `_alloc` for this interface, easier to read and more futur proof. But I am fine removing it.

dozylynx commented on vsock-argo/module/argo_ring.c:
```diff
@@ -0,0 +1,409 @@
+
+void argo_core_cleanup(void)
+{
+	unbind_from_irqhandler(argo_irq, NULL);
```
Should this unbind be conditional upon the bind having succeeded?

eric-ch replied:
`argo_core_init` would have failed which is a failure condition in `argo_transport_init`, so the module would not have been loaded if `bind_virq_to_irqhandler()` failed. A check on `argo_irq` is cheap though, so lets do it :).

dozylynx commented on vsock-argo/module/argo_ring.c:
```diff
@@ -0,0 +1,409 @@
+/*
+ * Initialisation and cleanup of the VIRQ.
+ */
+static int argo_irq = 11;
```
Not sure about this init with `11` here.

eric-ch replied:
Remnant of the former use as `argo_virq`. This should be `-1`.

dozylynx commented on vsock-argo/module/argo_ring.h:
```diff
@@ -0,0 +1,68 @@
+
+/*
+ * Messages on the ring are aligned on XEN_ARGO_MSG_SLOT_SIZE.
+ * Note: XEN_ARGO_MSG_SLOT_SIZE is a power of 2 right?
```
Yes it is. You could add a compile-time assert if you wanted to.

dozylynx commented on vsock-argo/module/argo_transport.c:
```diff
@@ -0,0 +1,557 @@
+
+	/* XXX: Assume skb is always in linear data area for now. */
+	mh = (void*)skb->data;
+	if (!mh)
```
maybe `pr_debug` warning here?

eric-ch commented:
### V4

- Removed duplicate function `argo_ring_has_data_nowrap`.
- Amended README.md intro with specific mention of tested platforms (x86-64).
- Moved `argo-compat.h` x86 specific header to `asm-x86` sub-directory and amend `KBuild` to include this architecture specific header only for X86 targets.
- Initialized `argo_irq` correctly and check successful VIRQ binding happened before attempting to unbind.
- Add debug print if the received `skb` data are not accessible.

dozylynx commented on vsock-argo/test/hatch/src/hatch.c:
```diff
@@ -0,0 +1,249 @@
+                perror("sendto");
+                goto out;
+            }
+            printd("%zuB sent to dom%u:%u.", rc, sa.svm_cid, sa.svm_port);
```
```
In file included from hatch.c:1:0:
hatch.c: In function ‘hatch_sendto_run’:
project.h:36:29: warning: format ‘%zu’ expects argument of type ‘size_t’, but argument 5 has type ‘int’ [-Wformat=]
             fprintf(stderr, "debug:%s:%d: " fmt "\n",   \
                             ^
hatch.c:131:13: note: in expansion of macro ‘printd’
             printd("%zuB sent to dom%u:%u.", rc, sa.svm_cid, sa.svm_port);
             ^~~~~~
hatch.c:131:23: note: format string is defined here
             printd("%zuB sent to dom%u:%u.", rc, sa.svm_cid, sa.svm_port);
                     ~~^
                     %u
```

dozylynx commented on vsock-argo/test/hatch/src/hatch.c:
```diff
@@ -0,0 +1,249 @@
+            case 'p':
+                rc = parse_uint(optarg, &args.local_port);
+                if (rc || !is_valid_port(args.local_port)) {
+                    printe("Invalid local port %lu.", args.local_port);
```
```
In file included from hatch.c:1:0:
hatch.c: In function ‘main’:
project.h:44:21: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘unsigned int’ [-Wformat=]
     fprintf(stderr, "error: " fmt "\n", ##__VA_ARGS__)
                     ^
hatch.c:220:21: note: in expansion of macro ‘printe’
                     printe("Invalid local port %lu.", args.local_port);
                     ^~~~~~
hatch.c:220:50: note: format string is defined here
                     printe("Invalid local port %lu.", args.local_port);
                                                ~~^
                                                %u
```

eric-ch commented:
### v5
- Amend embarrassing print format warnings.
- Use AM_FLAGS to pass `-W -Wall -Werror` by default to the C compiler.

dozylynx commented:
Builds clean now. Merging.
